### PR TITLE
[OTE_SDK] CVS-77228 Remove label exclusivity

### PR DESCRIPTION
**Pull request by Harmjandevries**

### CODE REVIEW

**Comment by Harmjandevries** [comment]
run ote_sdk tests

File: ote_sdk/ote_sdk/entities/label_schema.py [hunk 1]
```diff
@@ -540,9 +478,37 @@ def get_group_containing_label(self, label: LabelEntity) -> Optional[LabelGroup]
...
+        output += siblings
+
+        # Add all children of labels in the same group
+        if add_children:
+            for sibling in siblings:
+                output += self.get_children(sibling)
+
+        # Do the same for the parent of the label
+        if self.get_parent(label) is not None:
+            output += self.__get_exclusivity_recursion(self.get_parent(label), add_children=True)
```
Comment (atwinand) [inline ote_sdk/ote_sdk/entities/label_schema.py:510]: add_children = False

File: ote_sdk/ote_sdk/entities/label_schema.py [hunk 2]
```diff
@@ -540,9 +478,37 @@ def get_group_containing_label(self, label: LabelEntity) -> Optional[LabelGroup]
...
+    ) -> List[LabelEntity]:
+        """
+        Recursively computes all labels exclusive to a label. A label is exclusive with:
+        - All labels in the same group
+        - All children of labels in the same group
+        - All labels in the same group as any of the label's ancestors
+        - All children of labels in the same group as any of the label's ancestors
+
+        :param label: The label to get exclusive labels for
+        :return: List of labels exclusive to the label.
```
Comment (atwinand) [inline ote_sdk/ote_sdk/entities/label_schema.py:495]: docstring add_children

File: ote_sdk/ote_sdk/entities/label_schema.py [hunk 3]
```diff
@@ -540,9 +478,37 @@ def get_group_containing_label(self, label: LabelEntity) -> Optional[LabelGroup]
...
 
     def get_labels_exclusive_to(self, label: LabelEntity) -> List[LabelEntity]:
-        """Returns a list of labels that are exclusive to the passed label"""
-        self.__get_label(label)
-        return list(self.exclusivity_graph.neighbors(label))
+        """Returns a list of labels that are exclusive to the passed label."""
+        return self.__get_exclusivity_recursion(label, True)
+
+    def __get_exclusivity_recursion(
+        self, label, add_children: bool = False
```
Comment (atwinand) [inline ote_sdk/ote_sdk/entities/label_schema.py]: Since I was thinking of a wrong logic, maybe we can remove the add_children altogether.

Comment (Harmjandevries) [reply ote_sdk/ote_sdk/entities/label_schema.py]: Done

**Comment by Harmjandevries** [comment]
MR on SC will be merged before this: https://gitlab.devtools.intel.com/vmc-eip/IMPT/impt/-/merge_requests/5336

**Comment by druzhkov-paul** [comment]
@sovrasov FYI

**Comment by druzhkov-paul** [comment]
run ote_sdk tests

**Comment by Harmjandevries** [comment]
Discussed with Andru and we need to make some changes to that we can check for exclusivity of the empty label.

**Comment by Harmjandevries** [comment]
run ote_sdk tests

**Review by druzhkov-paul — DISMISSED** [review]

**Comment by Harmjandevries** [comment]
run ote_sdk tests

**Comment by Harmjandevries** [comment]
run ote_sdk tests

File: ote_sdk/ote_sdk/entities/label_schema.py [hunk 4]
```diff
@@ -503,17 +431,27 @@ def __append_group(self, label_group: LabelGroup):
 
     def are_exclusive(self, label1: LabelEntity, label2: LabelEntity) -> bool:
         """
-        Returns whether `label` and `label2` are mutually exclusive
+        Returns whether `label` and `label2` are mutually exclusive.
+        or if
```
Comment (atwinand) [inline ote_sdk/ote_sdk/entities/label_schema.py]: or if... ?

File: ote_sdk/ote_sdk/entities/label_schema.py [hunk 5]
```diff
@@ -540,9 +478,71 @@ def get_group_containing_label(self, label: LabelEntity) -> Optional[LabelGroup]
...
+            output += self.get_children(sibling)
+
+        # Do the same for the parent of the label
+        parent = self.get_parent(label)
+        if parent is not None:
+            output += self.__get_exclusivity_recursion(parent, add_empty=False)
+
+        # Add all empty labels that are not descendants of the label
+        if add_empty:
+            descendants = self.get_descendants(label)
```
Comment (atwinand) [inline ote_sdk/ote_sdk/entities/label_schema.py:521]: Descendants and ancestors, no?

Comment (Harmjandevries) [reply ote_sdk/ote_sdk/entities/label_schema.py:521]: This has the assumption built in that empty labels don't have descendants.

File: ote_sdk/ote_sdk/entities/label_schema.py (hunk 4 above)
Comment (Harmjandevries) [reply ote_sdk/ote_sdk/entities/label_schema.py]: Oops, that slipped through, thanks.

File: ote_sdk/ote_sdk/entities/label_schema.py (hunk 5 above)
Comment (atwinand) [reply ote_sdk/ote_sdk/entities/label_schema.py:521]: Fair assumption, please add as a comment

**Comment by Harmjandevries** [comment]
e2e passed: https://ci.iotg.sclab.intel.com/job/IMPT/job/SC-MR-dev/1379/

**Comment by druzhkov-paul** [comment]
run ote_sdk tests

**Review by druzhkov-paul — APPROVED** [review]

**Review by Ilya-Krylov — APPROVED** [review]

**Review by kprokofi — APPROVED** [review]

**Pull request merged**